server/src/test/kotlin/io/provenance/objectstore/gateway/server/ObjectStoreGatewayServerTest.kt

```diff
@@ -0,0 +1,101 @@
+import java.security.KeyPair
+import kotlin.random.Random
+import kotlin.test.assertContains
+
+class ObjectStoreGatewayServerTest {
+    lateinit var scopeFetchService: ScopeFetchService
+    lateinit var responseObserver: StreamObserver<GatewayOuterClass.FetchObjectResponse>
+    val keyPair: KeyPair = ProvenanceKeyGenerator.generateKeyPair()
+
+    lateinit var server: ObjectStoreGatewayServer
+
+    @BeforeEach
+    fun setUp() {
+        scopeFetchService = mockk()
+        responseObserver = mockk()
+
+        server = ObjectStoreGatewayServer(scopeFetchService)
+    }
+
+    @Test
+    fun `fetchObject should reject a request without an expiration time`() {
+        val request = keyPair.getValidRequest().toBuilder()
+            .apply {
+                paramsBuilder.clearExpiration()
+            }.build()
+
+        val exception = assertThrows<TimestampValidationException> {
+            server.fetchObject(request, responseObserver)
+        }
+
+        assertContains(exception.status.description!!, "Missing expiration", message = "Appropriate error should be thrown for missing expiration")
+    }
+
+    @Test
+    fun `fetchObject should reject a request with a past expiration time`() {
+        val request = keyPair.getValidRequest(-1000)
+
+        val exception = assertThrows<TimestampValidationException> {
+            server.fetchObject(request, responseObserver)
+        }
+
+        assertContains(exception.status.description!!, "Expired expiration", message = "Appropriate error should be thrown for past expiration")
+    }
+
+    @Test
+    fun `fetchObject should reject a request with an invalid signature`() {
+        val request = keyPair.getValidRequest().toBuilder()
+            .apply {
+                signatureBuilder.clearSignature()
+            }.build()
+
+        val exception = assertThrows<SignatureValidationException> {
+            server.fetchObject(request, responseObserver)
+        }
+
+        assertContains(exception.status.description!!, "Invalid signature", message = "Appropriate error should be thrown for invalid signature")
+    }
+
+    @Test
+    fun `fetchObject should succeed with a valid request`() {
+        val request = keyPair.getValidRequest()
+
+        val dummyRecords = listOf(GatewayOuterClass.Record.newBuilder()
+            .setName("dummyRecordName")
+            .addInputs(GatewayOuterClass.RawObject.newBuilder().setObjectBytes(Random.nextBytes(100).toByteString()))
+            .addOutputs(GatewayOuterClass.RawObject.newBuilder().setObjectBytes(Random.nextBytes(100).toByteString()))
+            .build())
+
+        every { scopeFetchService.fetchScope(request.params.scopeAddress, request.signature.signer.signingPublicKey.toPublicKey(), request.params.granterAddress.takeIf { it.isNotBlank() }) } returns dummyRecords
+
+        every { responseObserver.onNext(any()) } returns mockk()
+        every { responseObserver.onCompleted() } returns mockk()
```

Comment:
I haven't really used any mockk.  This syntax is really awesome, leveraging infix functions to create a pretty readable flow 👍